examples/coroutinesExample/src/main/java/io/realm/examples/coroutinesexample/ui/main/MainFragment.kt

```diff
@@ -21,76 +21,175 @@ import android.util.Log
-            viewModel.countDogs()
+    }
+
+    private fun setupRecyclerView() {
+        with(binding.list) {
+            layoutManager = LinearLayoutManager(context)
+            adapter = newsReaderAdapter
         }
-        view.findViewById<Button>(R.id.buttonDelete).setOnClickListener {
-            // Calling this while bulk-inserting will not freeze the UI
-            viewModel.deleteAll()
+
+        with(binding.refresh) {
+            setOnRefreshListener {
+                with(binding.spinner) {
+                    viewModel.getTopStories(getApiSection(adapter, selectedItemPosition), true)
+                }
+            }
         }
-        view.findViewById<Button>(R.id.buttonCancelCoroutine).setOnClickListener {
-            viewModel.cancel()
+    }
+
+    private fun setupLiveData() {
+        viewModel.newsReaderState.observe(viewLifecycleOwner, Observer { viewState ->
+            when (viewState) {
+                is NewsReaderState.Loading -> {
+                    Log.d(TAG, "--- origin: ${viewState.origin}, loading")
+                    RealmStateHelper.loading(binding)
+                }
+                is NewsReaderState.Data -> {
+                    Log.d(TAG, "--- origin: ${viewState.origin}, elements: ${viewState.data.size}")
+                    RealmStateHelper.data(binding, viewState.data, newsReaderAdapter)
+                }
+                is NewsReaderState.NoNewData -> {
+                    Log.d(TAG, "--- origin: ${viewState.origin}, no new data")
+                    RealmStateHelper.noNewData(binding)
+                }
+                is NewsReaderState.ErrorException -> {
+                    val stacktrace = viewState.throwable.cause?.stackTrace?.joinToString { "$it\n" }
+                    Log.e(TAG, "--- error (exception): ${viewState.throwable.message} - ${viewState.throwable.cause?.message}: $stacktrace")
+                    RealmStateHelper.error(binding)
+                }
+                is NewsReaderState.ErrorMessage -> {
+                    Log.e(TAG, "--- error (message): ${viewState.message}")
+                    RealmStateHelper.error(binding)
+                }
+            }
+        })
+    }
+
+    private fun getApiSection(adapter: SpinnerAdapter, position: Int): String {
+        val apiSection = adapter.getItem(position) as String
+        return requireNotNull(sectionsToNames[apiSection])
+    }
+
+    companion object {
+
+        const val TAG = "MainFragment"
+
+        fun newInstance() = MainFragment()
+    }
+}
+
+sealed class NewsReaderState {
+
+    abstract val origin: String
+
+    data class Loading(override val origin: String) : NewsReaderState()
+    data class Data(override val origin: String, val data: List<RealmNYTimesArticle>) : NewsReaderState()
+    data class NoNewData(override val origin: String) : NewsReaderState()
+    data class ErrorException(override val origin: String, val throwable: Throwable) : NewsReaderState()
```

Comment:
Any reason we need both `ErrorException` and `ErrorMessage` as states? Doesn't that just complicate things since we are just reading the error message from the exception anyway?

Reply:
IIRC this is because Store can emit either an "error exception" or an "error message". I decided to leave it the same way for consistency.